File: app/controllers/api/v2/subscriptions_controller.rb
#
# Copyright 2013 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.

class Api::V2::SubscriptionsController < Api::V1::SubscriptionsController

  include Api::V2::Rendering

  resource_description do
    description "Systems subscriptions management."
    param :system_id, :identifier, :desc => "System uuid", :required => true

    api_version 'v2'
  end

  api :POST, "/systems/:system_id/subscriptions", "Create a subscription"
  param :subscription, Hash, :required => true, :action_aware => true do
    param :pool, String, :desc => "Subscription Pool uuid", :required => true
    param :quantity, :number, :desc => "Number of subscription to use", :required => true
  end
  def create
    expected_params = params[:subscription].with_indifferent_access.slice(:pool, :quantity)
    raise HttpErrors::BadRequest, _("Please provide pool and quantity") if expected_params.count != 2
    @system.subscribe(expected_params[:pool], expected_params[:quantity])
    respond :resource => @system
  end

  def index
    subscriptions = {
        :subscriptions => @system.consumed_entitlements,
        :subtotal => @system.consumed_entitlements.count,
        :total => @system.consumed_entitlements.count
    }

    respond({:collection => subscriptions})
  end

end

File: engines/bastion/app/assets/javascripts/bastion/subscriptions/subscriptions.controller.js
/**
 * Copyright 2013-2014 Red Hat, Inc.
 *
 * This software is licensed to you under the GNU General Public
 * License as published by the Free Software Foundation; either version
 * 2 of the License (GPLv2) or (at your option) any later version.
 * There is NO WARRANTY for this software, express or implied,
 * including the implied warranties of MERCHANTABILITY,
 * NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
 * have received a copy of GPLv2 along with this software; if not, see
 * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
*/

/**
 * @ngdoc object
 * @name  Bastion.subscriptions.controller:SubscriptionsController
 *
 * @requires $scope
 * @requires $q
 * @requires $location
 * @requires gettext
 * @requires Nutupane
 * @requires Subscription
 * @requires Provider
 * @requires CurrentOrganization
 * @requires unlimitedFilterFilter
 *
 * @description
 *   Provides the functionality specific to Subscriptions for use with the Nutupane UI pattern.
 *   Defines the columns to display and the transform function for how to generate each row
 *   within the table.
 */
angular.module('Bastion.subscriptions').controller('SubscriptionsController',
    ['$scope', '$filter', '$q', '$location', 'gettext', 'Nutupane', 'Subscription', 'Provider', 'CurrentOrganization', 'SubscriptionsHelper',
    function ($scope, $filter, $q, $location, gettext, Nutupane, Subscription, Provider, CurrentOrganization, SubscriptionsHelper) {

        var params = {
            'organization_id':  CurrentOrganization,
            'search':           $location.search().search || "",
            'sort_by':          'name',
            'sort_order':       'ASC',
            'enabled' :         true,
            'paged':            true
        };

        var nutupane = new Nutupane(Subscription, params);
        $scope.table = nutupane.table;
        $scope.refreshTable = nutupane.refresh;
        $scope.successMessages = [];
        $scope.errorMessages = [];

        $scope.table.closeItem = function () {
            $scope.transitionTo('subscriptions.index');
        };

        $scope.groupedSubscriptions = {};
        $scope.$watch('table.rows', function (rows) {
            $scope.groupedSubscriptions = SubscriptionsHelper.groupByProductName(rows);
        });

        $scope.formatConsumed = function (subscription) {
            var quantity = $filter('unlimitedFilter')(subscription.quantity);

            return gettext('%(consumed)s out of %(quantity)s').replace('%(consumed)s', subscription.consumed).replace('%(quantity)s', quantity);
        };

        $scope.formatInstanceBased = function (subscription) {
            if (subscription['instance_multiplier'] === undefined || subscription['instance_multiplier'] === "" || subscription['instance_multiplier'] === 0) {
                return gettext("No");
            }
            return gettext("Yes");
        };

        $scope.subscriptionType = function (subscription) {
            if (subscription['virt_only']) {
                return gettext('Virtual');
            } else {
                return gettext('Physical');
            }
        };

        $scope.providers = Provider.query({ 'provider_type': 'Red Hat' }, function (response) {
            $scope.provider = _.first(response.results);
        });

        $scope.subscriptions = Subscription.query();

        $q.all([$scope.providers.$promise, $scope.subscriptions.$promise]).then(function () {
            if ($scope.subscriptions.results.length < 1) {
                $scope.transitionTo('subscriptions.manifest.import', {providerId: $scope.provider.id});
            }
        });
    }]
);
